controller/src/run.rs

```diff
@@ -62,38 +56,45 @@ async fn controller_main() -> NeverResult {
 
     tracing_handle.attach_nats(nats.clone())?;
 
-    let mut futs: Vec<Pin<Box<dyn Future<Output = NeverResult>>>> = vec![];
+    let _scheduler_supervisors = if scheduler_plan.is_some() {
+        let scheduler = {
+            let nats = nats.clone();
+            let state = state.clone();
+            Supervisor::new("scheduler", move || {
+                run_scheduler(nats.clone(), state.clone())
+            })
+        };
 
-    if scheduler_plan.is_some() {
-        futs.push(Box::pin(run_scheduler(nats.clone(), state.clone())));
+        let backend_state = {
+            let nats = nats.clone();
+            Supervisor::new("backend_state", move || {
+                update_backend_state_loop(nats.clone())
+            })
+        };
 
-        futs.push(Box::pin(update_backend_state_loop(nats.clone())));
+        let monitor_drone_state = {
+            let nats = nats.clone();
+            Supervisor::new("monitor_drone_state", move || {
+                monitor_drone_state(nats.clone())
+            })
+        };
 
-        futs.push(Box::pin(monitor_drone_state(nats.clone())));
-    }
+        Some((scheduler, backend_state, monitor_drone_state))
+    } else {
+        None
+    };
 
-    if let Some(dns_plan) = dns_plan {
-        futs.push(Box::pin(serve_dns(dns_plan)));
-    }
+    #[allow(clippy::manual_map)]
+    let _dns_supervisor = if let Some(dns_plan) = dns_plan {
+        Some(Supervisor::new("dns", move || serve_dns(dns_plan.clone())))
+    } else {
+        None
+    };
 
-    try_join_all(futs.into_iter()).await?;
-    // try_join_all either returns an Err, or Ok() with a list of Never values.
-    // Since Never values are not constructable, if we get here, we can assume that
-    // try_join_all returned an empty list. This means that no event loops ever
-    // actually ran, i.e. futs was empty.
-    Err(anyhow!("No event loops selected."))
+    wait_for_interrupt().await
```

Comment:
cool! so when this goes off it'll actually do the drop for the event loops right? does this mean containers get shut down?

Reply:
it will drop the event loops, but it won't terminate the containers.